compiler/lcalc/compile_without_exceptions.ml

```diff
@@ -0,0 +1,455 @@
+  | D.LEmptyError ->
+      Errors.raise_spanned_error
+        "Internal Error: An empty error was found in a place that shouldn't be possible." pos
+
+(** [c = disjoint_union_maps cs] Compute the disjoint union of multiple maps. Raises an internal
+    error if there is two identicals keys in differnts parts. *)
+let disjoint_union_maps (pos : Pos.t) (cs : 'a A.VarMap.t list) : 'a A.VarMap.t =
+  let disjoint_union =
+    A.VarMap.union (fun _ _ _ ->
+        Errors.raise_spanned_error
+          "Internal Error: Two supposed to be disjoints maps have one shared key." pos)
+  in
+
+  List.fold_left disjoint_union A.VarMap.empty cs
+
+(** [e' = translate_and_cut ctx e ] Translate the Dcalc expression e into an expression in Lcalc,
+    given we translate each cuts correctly. It ensures the equivalence between the execution of e
+    and the execution of e' are equivalent in an environement where each variable v, where (v, e_v)
+    is in cuts, has the non-empty value in e_v. *)
+let rec translate_and_cut (ctx : ctx) (e : D.expr Pos.marked) : A.expr Pos.marked Bindlib.box * cuts
+    =
+  let pos = Pos.get_position e in
+  match Pos.unmark e with
+  (* empty-producing/using terms. We cut those. (D.EVar in some cases, EApp(D.EVar _, [ELit LUnit]), EDefault _, ELit LEmptyDefault) I'm unsure about assert. *)
+  | D.EVar v ->
+      (* todo: for now, we requires there is unpure variables. This can change if the said variable are always present in the tree as thunked. *)
+      let v, pos_v = v in
+      if not (find ~info:"search for a variable" v ctx).is_pure then begin
+        let v' = A.Var.make (Bindlib.name_of v, pos_v) in
+        Cli.debug_print
+        @@ Format.asprintf "Found an unpure variable %a, created a variable %a to replace it" pp_var
+             v pp_var v';
+        (A.make_var (v', pos), A.VarMap.singleton v' e)
+      end
+      else ((find ~info:"should never happend" v ctx).expr, A.VarMap.empty)
+  | D.EApp ((D.EVar (v, pos_v), p), [ (D.ELit D.LUnit, _) ]) ->
+      if not (find ~info:"search for a variable" v ctx).is_pure then begin
+        let v' = A.Var.make (Bindlib.name_of v, pos_v) in
+        Cli.debug_print
+        @@ Format.asprintf "Found an unpure variable %a, created a variable %a to replace it" pp_var
+             v pp_var v';
+        (A.make_var (v', pos), A.VarMap.singleton v' (D.EVar (v, pos_v), p))
+      end
+      else
+        Errors.raise_spanned_error
+          "Internal error: an pure variable was found in an unpure environment." pos
+  | D.EDefault (_exceptions, _just, _cons) ->
+      let v' = A.Var.make ("default_term", pos) in
+      (A.make_var (v', pos), A.VarMap.singleton v' e)
+  | D.ELit D.LEmptyError ->
+      let v' = A.Var.make ("empty_litteral", pos) in
+      (A.make_var (v', pos), A.VarMap.singleton v' e)
+  | D.EAssert _ ->
+      (* as discuted, if the value in an assertion is empty, an error should the raised. This beavior is different from the ICFP paper. *)
```

Comment:
To keep things consistent with the ICFP paper, a simple alternative is to wrap all assertions' contents with `error_on_empty` during the Scopelang to Dcalc translation.

Reply:
I prefer to fix this after the merge then.

Reply:
Ok but at least put a TODO

Reply:
fixed